internal/cmd/agent.go

```diff
@@ -0,0 +1,259 @@
+package cmd
+
+import (
+	"cdr.dev/coder-cli/internal/x/xcobra"
+	"cdr.dev/coder-cli/internal/x/xwebrtc"
+	"cdr.dev/coder-cli/pkg/proto"
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/sloghuman"
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/hashicorp/yamux"
+	"github.com/pion/webrtc/v3"
+	"github.com/spf13/cobra"
+	"golang.org/x/xerrors"
+	"io"
+	"net"
+	"net/url"
+	"nhooyr.io/websocket"
+	"os"
+	"strings"
+	"time"
+)
+
+func agentCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:    "agent",
+		Short:  "Run the workspace agent",
+		Long:   "Connect to Coder and start running a p2p agent",
+		Hidden: true,
+	}
+
+	cmd.AddCommand(
+		startCmd(),
+	)
+	return cmd
+}
+
+func startCmd() *cobra.Command {
+	var (
+		token    string
+	)
+	cmd := &cobra.Command{
+		Use:   "start [coderURL] --token=[token]",
+		Args:  xcobra.ExactArgs(1),
+		Short: "starts the coder agent",
+		Long:  "starts the coder agent",
+		Example: `# start the agent and connect with a Coder agent token
+
+coder agent start https://my-coder.com --token xxxx-xxxx
+
+# start the agent and use CODER_AGENT_TOKEN env var for auth token
+
+coder agent start https://my-coder.com
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := cmd.Context()
+			log := slog.Make(sloghuman.Sink(cmd.OutOrStdout()))
+
+			// Pull the URL from the args and do some sanity check.
+			rawURL := args[0]
+			if rawURL == "" || !strings.HasPrefix(rawURL, "http") {
+				return xerrors.Errorf("invalid URL")
+			}
+			u, err := url.Parse(rawURL)
+			if err != nil {
+				return xerrors.Errorf("parse url: %w", err)
+			}
+			// Remove the trailing '/' if any.
+			u.Path = "/api/private/envagent/listen"
+
+			if token == "" {
+				var ok bool
+				token, ok = os.LookupEnv("CODER_AGENT_TOKEN")
+				if !ok {
+					return xerrors.New("must pass --token or set the CODER_AGENT_TOKEN env variable")
+				}
+			}
+
+			q := u.Query()
+			q.Set("agent_token", token)
+			u.RawQuery = q.Encode()
+
+			ctx, cancelFunc := context.WithTimeout(context.Background(), time.Second*15)
+			defer cancelFunc()
+			log.Info(ctx, "connecting to broker", slog.F("url", u.String()))
+			conn, _, err := websocket.Dial(ctx, u.String(), nil)
+			if err != nil {
+				return fmt.Errorf("dial: %w", err)
+			}
+			nc := websocket.NetConn(context.Background(), conn, websocket.MessageBinary)
```

Comment:
I know this is an MVP but there's a lot going on in this command so all of this should move into it's own package (or potentially even it's own repo if we want to use it within our monorepo as well).

Reply:
I pinky promise I'll do that if this ends up being a thing we stick with ;)